### PR TITLE
Allow default route dependencies

### PR DESCRIPTION
**Pull request by rhysrevans3**
**Description:**
Allows for `*` to be used for the `path` or `method` variables in a route dependencies' scope to match all allowable routes.

**PR Checklist:**

- [x] `pre-commit` hooks pass locally
- [x] Tests pass (run `make test`)
- [x] Documentation has been updated to reflect changes, if applicable, and docs build successfully (run `make docs`)
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/stac-fastapi/blob/main/CHANGES.md).

### CODE REVIEW

File: stac_fastapi/api/stac_fastapi/api/routes.py [hunk 1]
```diff
@@ -100,28 +101,39 @@ def add_route_dependencies(
...
-                    ),
-                )
-
-            # Register dependencies directly on route so that they aren't ignored if
-            # the routes are later associated with an app (e.g.
-            # app.include_router(router))
-            # https://github.com/tiangolo/fastapi/blob/58ab733f19846b4875c5b79bfb1f4d1cb7f4823f/fastapi/applications.py#L337-L360
-            # https://github.com/tiangolo/fastapi/blob/58ab733f19846b4875c5b79bfb1f4d1cb7f4823f/fastapi/routing.py#L677-L678
-            route.dependencies.extend(dependencies)
+            if hasattr(route, "dependant"):
```
Comment (rhysrevans3) [inline stac_fastapi/api/stac_fastapi/api/routes.py:123]: This is needed for paths `/api`, `/api.html', `/docs/oauth2-redirect` as they don't have the dependant attribute.

File: stac_fastapi/api/stac_fastapi/api/routes.py [hunk 2]
```diff
@@ -100,28 +101,38 @@ def add_route_dependencies(
...
     """
     for scope in scopes:
+        _scope = copy.deepcopy(scope)
         for route in routes:
-            match, _ = route.matches({"type": "http", **scope})
+            if scope["path"] == "*":
+                _scope["path"] = route.path
+
+            if scope["method"] == "*":
+                _scope["method"] = list(route.methods)[0]
```
Comment (jonhealy1) [inline stac_fastapi/api/stac_fastapi/api/routes.py:116]: Is this correct or should we apply dependencies to each method individually?

Comment (rhysrevans3) [reply stac_fastapi/api/stac_fastapi/api/routes.py:116]: I believe so, the method is only used to get a full match with the route [here](https://github.com/stac-utils/stac-fastapi/pull/705/files#diff-3f814cfc24182540359cf16e8a3cedff38803d2feb6e489f7fc83b2e46758287R118-R120). The dependency is then applied to the route so is not specific to the method.

**Comment by vincentsarago** [comment]
@alukach would you be able to review this one? 🙏

**Review by jonhealy1 — APPROVED** [review]
LGTM nice work!

File: stac_fastapi/api/stac_fastapi/api/routes.py (hunk 1 above)
Comment (alukach) [reply stac_fastapi/api/stac_fastapi/api/routes.py:122]: If we just use a guard here, we can avoid indenting the remaining code.
```suggestion
            # Ignore paths without dependants, e.g. /api, /api.html, /docs/oauth2-redirect
            if not hasattr(route, "dependant"):
                continue
```

**Review by alukach — APPROVED** [review]
This is a nice convenience feature.  I think it makes sense. At first I thought a user should instead just do something like:

```py
routes.add_route_dependencies(
    app.router.routes,
    [
        {"path": route.path, "method": method, "type": "http"}
        for route in app.router.routes
        for method in route.methods
    ],
    [Depends(auth.basic_auth)],
)
```

However, the value of being able to something like this seems really enticing:

```py
routes.add_route_dependencies(
    app.router.routes,
    [
        {"path": "*", "method": method, "type": "http"}
        for method in ("POST", "PUT", "DELETE")
    ],
    [Depends(auth.basic_auth)],
)
```

Left a minor code style suggestion, but 👍 💯

**Review by vincentsarago — APPROVED** [review]

**Pull request merged**